Support 608/708 in H262 + parse isWideAspectRatio flag

peddisri opened the pull request:
Major re-design of CEA708 Cues decoder and subtitle rendering to make it compliant with Sarnoff test streams. Added support for
missing features required to pass Sarnoff tests. Minor fixes in CEA608 decoder too.

peddisri commented:
@ojw28 Hi. Can you review this please?

ojw28 commented:
Please can you provide a more detailed change description, so we don't have to reverse engineer what the description should be from the code :)? Please be specific about where the test streams are, how we can run the tests, and what the "missing features" are. Thanks!

peddisri commented:
Sarnoff compliance test streams are to be licensed. https://www.sri.com/engage/products-solutions/sarnoff-video-test-tools 
 So I can't share it here. They have exhaustive test streams to test CEA 708 spec compatibility.  The original code failed many tests. And I think FCC compliance guidelines requires compatibility with Sarnoff tests

1. Firstly, the streams are MPEG2. So we had to add support to parse CC data out of it. Hence the changes in H262Reader.java and added UserDataReader.java file.

2. Sometimes the CC text is all jumbled up for H.264 content.

[Root Cause] 
There is a timing issue in rendering CC.
Due to H264 B frames decoding and presentation order mismatch, sometimes,
older timestamps CC frames are being decoded that resulted into
jumbled CC text.

This happens due to several reasons.
a. Input buffers are not pre-buffered. This does not allow input
buffers to be pre-parsed and generate subtiles ready for rendering.
Sometimes, input frames timestamps are more than the current playback
position. So by allowing pre-buffering we hope this situation does not
occur.

b. There is no direct rate control mechanism at input buffers. It indirectly
depends on output buffers, but that depends on the content.
i.e depends on how many input buffers constitue one renderable cue.

c.. Input buffers whose timestamp is greater than the last decoded buffers
are being queued to Decoder.

[Solution]
a. We reject input frames that are older than the last decoded frames.
b. We do not start processing input frames till we have received at least
   half of the input frames that can be held in the input queue.

3. Added support for HCR and FF command.  Test using C67GSETS.TRP of Sarnoff test streams

4. To support DLY and DLC commands, we need to have a service input buffer.  Also the current design of CEA708 Decoder is not according to the spec. As per the spec CCP and service blocks are asynchronous. So we can't wait to start parsing service blocks until after generating a complete CCP. The parsing of various layers much be chained and be executed in parallel.

[Solution] We removed the DTVCC packet usage and added functions to parse
incoming data inline. We create a service input buffer which is a circular buffer
that stores all the incoming CC data.


5.Fix for handling multiple service blocks in one CCP
As per the spec, a CCP packet can contain more than one Service Blocks with same service number. Current implementation does not handle this.

 6. Styling such as color, size, location, font, superscript, subscript, edge style etc were not supported. These features were added.  

7. ,  Curently, CEA708Decoder is not aware of aspect ratio. The spec has some differences between the way CC is rendered in 4:3 and 16:9 content, specifically about number of columns it can have 

8. Black window color not rendered.
[Problem] When the window fill color is black, it is not rendered.
The window is transparent.
[Root Cause] Default implementation ignores the window color when it
is set to black. Instead it picks the color from CaptioningManager.
Unfortunately, the default value of window color is set to transparent in captioning manager.  

[Solution] Even if the window color is set to black, explicitly flag the subtitle painter to fill the window with black color. This will make sure that the default window color is drawn black.

9. Adding Parity bit and Valid bit check for Closed Captions. We failed C40UDTAb test file of Sarnoff. 

10. Support for column count and column lock added. 

11. Fixes in supporting user configured styling in captioning manager. 
[Problem]
a. By default, embedded styling is preferred over user styling. But as per the spec, user styling should be preferred over embedded styling.
b. User styling should not override italics, underline and offsets of embedded styling.There is no user style setting for these parameters.

[Root cause]
a. Application of Embedded styling is enabled by default.
b. If user styling is applied, all overridden spans are removed.

**12.Fixed Window Map issues**
Clear Window tests failed. Clearing window removes the window from display, it should just clear the text..
Test Content: c68winda.trp. Video Id:30 and CC Id: eng,id:30

2. CC not displayed. Pen style 7 test failed.
Test Content: c68winda.trp. Video Id: 32 and CC id por, id:32

[Root cause]
1. Currently, clear window clears the cues and we don't allow empty cues to be rendered.
The spec however allows rendering empty window.

2. Extended Service number should not be added to the service number
read from the standard service block header.
Instead, it should be read from the extended service block header.

[Solution]
a. We allow rendering empty captions. Additionally, we calculate the window width independent of the text width by using max column count set in Define Window Command (DFW) command and the width of the font. 

b. Extended service number is NOT added to service number, but just read.

c. One minor fix in Window Style Justification. Incorrectly set to
left instead of center.

[TEST]
   Verification tests of c68winda.trp of Sarnoff streams
   a. Video Id:30 and CC Id: eng,id:30.
   b. Video Id: 32 and CC id por, id:32

13. Added support for top to bottom scroll direction

14. Honor process_cc_data_flag - i.e skip if not set.

15. For CEA608, support Paint ON caption mode

16. To support SetPenLocation (row and column) we need to modify the Cue structure and the way we paint cues in subtitlepainter. This required a major re-design in this area.  If we just insert spaces in the text to start writing from specified column, we will have black background due to "space" character. Ideally, if the window background color is transparent, it should not render anything between 0 to column offset.  Some sarnoff tests failed due to this.

17. If we start decoding CC in between the live content, and the defineWindow command was in the past, we should not render Cues to a window until the window is defined. Rendering cues on undefined window causes unintended CC displays

**18.  Cea608 channels are not handled properly**
[Problem]
CC1 is shown even when CC2-3-4 are selected

[Solution]
There are multiple issues present:
* Incoming channel and frame information is not saved in Cea608Decoder
* Misc Code handling does not use Frame information
* Control code handling has TODOs for multiple channels

[Test]
Sarnoff test C32FR29b.trp:
* has CC1 and CC3 must be rendered with corresponding texts
* CC2 and CC4 must not show up

**19. Handle different video & caption aspect ratio case**

[Problem] If the captions was authored for a aspect ratio that is different than the video aspect ratio, we ignore that and instead consider video aspect ratio to render the captions. This use case typically happens if the video is 16:9 ( wide aspect ratio ) but the captions are authored considering 4:3 content.

[Solution]
Pass the caption aspect ratio to subtitleview and there adjust the width of the cue box if the caption aspect ratio is 4:3 but video aspect ratio is 16:9

[Tests] CS3 caption of C37SZE1 in Sarnoff test suite

**20. Other CEA608 Fixes:**

   -- ERASE command is not processed immediately
   -- Midrow code whitespaces. We have missing whitespaces when italic or other mid-row formattingis used

21. CEA608 NTSC deleted tests
Cea608 has dedicated bandwidth, the bytes should always be present in the stream - at least invalid or null bytes. If they are not present the stream is probably modified and the previously recevied data (already on the screen) is invalid. Sarnoff tests expect the screen being cleared

[Solution]
Clear the screen when there is no 608 data in the stream. It should be a rare
event

[Test]
* run c40udtab: UDTESTS. Check the "NTSC Deleted test"
* run the various frame rate tests: c30-c35

22. CEA608 - Alignment issues

[Problem]
* Switching to Paint-on style is not reflected immediately in the CueBuilders * We right align captions immediately when there is more whitespace on the left side of the text than on the right side

[Solution]
* update CueBuilders immediately when captionMode changes
* Do not right align text if there is more then 3 characters could be added
to the same row.

[Test]
* Check lots of live streams.
* Sarnoff tests should probably not show any difference, these are rare use cases
* Beginning of commercials tend to have paint on captions (as there are no video
frames shown before the beginning of the commercials that could transport the
caption data to pre-load them. They want to show the characters as soon as they
arrive - paint on mode is doing that.
  
22. Window Size fixes:
Window size depends on max column and effective font size, which depends on embedded font scale.

peddisri commented:
Features not implemented and known issues:

1. Other Scroll and Print directions
2. Window Style 7
3. Full justification
4.  c68winda.trp : Video Track ID 32 and CC Track index 8 and 9 does not render text because of the pen style has uniform black edge. The EdgeStyleSpan sets  the color of the pen as black which also changes the text color. This is a limitation of the way span works. We can't implement uniform style correctly without causing this regression.  The only way is to implement a ReplaceamentSpan, which means the derived class needs to implement all style features. Too complicated !

peddisri commented:
@ojw28  is this information sufficient?

ojw28 commented:
Yes, although when fixing a lot of separate issues it's *much* easier to review the changes if they're split into separate commits (one for each issue, with a corresponding commit message that accurately describes it). And perhaps separate pull requests if it's likely some parts of the change are less controversial and/or simpler than others. Handling a single large commit requires the reviewer to manually untangle it to figure out which parts correspond to which feature or fix.

We will look at this when we have a chance. It's likely we'll look at parts of it far more quickly if it's split into smaller changes. Up to you whether to split or wait :).

ojw28 commented:
Looking at this a bit more, it would be really helpful if it could be split up into smaller pieces. For example could we have a pull request that just fixes https://github.com/google/ExoPlayer/issues/3906, and one that just fixes https://github.com/google/ExoPlayer/issues/4321? I'd expect the first to be very small, and the second to be a bit larger. Then we can continue picking off medium size pieces until we get everything landed.

peddisri commented:
It took many patches to get to this state. There were incremental improvements during development, i.e some change that happened to fix an issue was later reworked. And there are too many patches. So it would be convenient if all of them are reviewed under one patch and pull request. During review, feel free to ask clarification about a specific change and I'll be happy to answer. Most of the change is self explanatory.

ojw28 commented:
I don't think a 2500 line diff that corresponds to 22 features/issues, collapsed together, can ever be classed as self explanatory :). Or easy to review for that matter. Sending the change as a single pull request also means that we can't fix any of the 22 features/issues until we're happy with everything, which isn't ideal.

I think we can do one of two things:

1. You split the change up into more manageable pieces, preferably with the easy bits first.
1. We'll merge the whole change into a new branch, then do the work of teasing out the individual features/issues ourselves, merging them into `dev-v2` until we've ticked everything off.

Please let us know which option suits you best. I'm happy to have a stab at the second approach, if you prefer, in which case no action is required from your side. Although for future changes please keep in mind that it's much easier to handle pull requests that do not conflate many individual changes together.

Thanks!

ojw28 commented:
p.s. We are looking into getting hold of the test streams for internal testing going forward.

ojw28 commented on library/ui/src/main/java/com/google/android/exoplayer2/ui/SubtitleView.java:
```diff
@@ -70,8 +71,10 @@ public SubtitleView(Context context, AttributeSet attrs) {
     textSize = DEFAULT_TEXT_SIZE_FRACTION;
-    applyEmbeddedStyles = true;
-    applyEmbeddedFontSizes = true;
+    // As per the CEA specs user styling should take priority over embedded styling.
```
This view is not CEA specific, so it shouldn't be defaulting to whatever happens to be correct for CEA. If you know you're dealing with 608 captions then you can call the corresponding methods to achieve the same thing. Alternatively, the preference would need baking into the cues themselves, or equivalent.

I think it's probably best just to let the app configure things?

peddisri replied:
Well, the basic principle behind this recommendation from CEA is that the user preference has higher priority than content provider. I'm okay if this configuration is done by app.

ojw28 commented on library/ui/src/main/java/com/google/android/exoplayer2/ui/SubtitleView.java:
```diff
@@ -84,17 +87,18 @@ public void onCues(List<Cue> cues) {
    * Sets the cues to be displayed by the view.
    *
-   * @param cues The cues to display, or null to clear the cues.
+   * @param cues The cues to display.
```
Revert? (and restore the nullable annotation).

peddisri replied:
Ok.

ojw28 commented on library/ui/src/main/java/com/google/android/exoplayer2/ui/SubtitleView.java:
```diff
@@ -126,10 +130,15 @@ public void setFixedTextSize(int unit, float size) {
    */
   public void setUserDefaultTextSize() {
-    float fontScale = Util.SDK_INT >= 19 && !isInEditMode() ? getUserCaptionFontScaleV19() : 1f;
+    fontScale = Util.SDK_INT >= 19 && !isInEditMode() ?
```
I'm not sure I understand what's happening here. Doesn't fontScale end up being applied twice (via setFractionalTextSize and when passed directly to painter.draw)?

peddisri replied:
The fontScale passed to painter.draw is only used to decide the effectiveFontScale that is required to calculate the window size. It is not directly used in the calculation of effectiveFontScale, but to decide what it should be

ojw28 commented on library/ui/src/main/java/com/google/android/exoplayer2/ui/SubtitleView.java:
```diff
@@ -187,7 +196,7 @@ public void setApplyEmbeddedStyles(boolean applyEmbeddedStyles) {
 
   /**
-   * Sets whether font sizes embedded within the cues should be applied. Enabled by default.
+   * Sets whether font sizes embedded within the cues should be applied. Disabled by default.
```
Revert

ojw28 commented on library/ui/src/main/java/com/google/android/exoplayer2/ui/SubtitlePainter.java:
```diff
@@ -150,19 +289,88 @@ public SubtitlePainter(Context context) {
     boolean isTextCue = cue.bitmap == null;
     int windowColor = Color.BLACK;
+    int windowBorderColor = Color.BLACK;
+    // Spec: 8.4.6 Window Size
```
This class is not CEA specific, and shouldn't contain CEA specific comments or code.

peddisri replied:
The concept of Window May be unique to CEA, but it can be generically implemented in SubtitlePainter, and other caption types can set the default values. 

I can remove the CEA specific  comment though.

ojw28 replied:
* I don't think the concept of a window is unique to CEA
* If there are any assumptions around the font being monospace in the code below, then I don't think it's generic enough. I think we'd need something more like a `Cue.sizeType` variable that can indicate `Cue.size` is something other than a fraction.

peddisri commented:
I prefer the 2nd option too.  The thing is I had some 25-30 changes, and It took quite an effort to squash them into one for this pull request. (up-merge to 2.8.1 , remove our change markers in the code etc). 

I do understand and agree that incremental changes are easier to handle and review. But this specific one is like  a complete re-write./re-design and I felt it would be a better approach to review it in entirety. 

In future it'll be incremental specific changes, for sure.

ojw28 commented on library/core/src/main/java/com/google/android/exoplayer2/text/CaptionStyleCompat.java:
```diff
@@ -67,7 +67,9 @@
   public static final CaptionStyleCompat DEFAULT = new CaptionStyleCompat(
-      Color.WHITE, Color.BLACK, Color.TRANSPARENT, EDGE_TYPE_NONE, Color.WHITE, null);
+          Color.WHITE, Color.BLACK,
+          /* Transparent Blue */Color.BLUE & (0x00FFFFFF),
```
Why has this been changed to transparent blue? Note this class is not CEA specific. Ditto for changing the default to MONOSPACE below.

peddisri replied:
The Default Window Color from caption manager in some devices that i tested was this color. If we need to apply user styling, we need to know if the user has modified this value from the default or not. So , we need to change the default value.  LIne 325 of SubtitlePainer.java
I'm not sure if this is a bug in some devices or in AOSP code. i.e instead of opaque black. its  transparent. 

As for Monospace, the calculation of Window Size (specifically width) is a function of max window columns and font size. i.e window width = maxWindowColumns * fontSize;
Now if the text is not monochrome, the calculated window size and the actual text size results into a visual where window is disproportionately larger or smaller than the text. By using MONOSPACE the window size and text size are calculated correctly and some of the sarnoff tests visual output is comparable to the reference output.

ojw28 replied:
As above, this class is not CEA specific. You can't change the default to MONOSPACE just because it fixes CEA-608, because it will affect how other caption types are rendered. If we need to be able to specify that CEA-608 uses monospace fonts, then that information needs to be hardcoded somewhere in CEA-608 specific code. For example by setting styling information on each `Cue`. The text package then needs to be able to handle that information in the correct way.

ojw28 commented on library/core/src/main/java/com/google/android/exoplayer2/extractor/ts/DefaultTsPayloadReaderFactory.java:
```diff
@@ -137,8 +141,34 @@ public TsPayloadReader createPayloadReader(int streamType, EsInfo esInfo) {
+   * @param esInfo The {@link EsInfo} passed to {@link #createPayloadReader(int, EsInfo)}.
+   * @return A {@link List<Format>} containing list of closed caption formats.
+   */
+  private List<Format>  getCCformats(EsInfo esInfo) {
```
This method appears to still be skipping the wide_aspect_ratio bit. In practice I think this means `Format.isWideAspectRatio` is being set to true along all code paths, which means that it's not currently doing anything useful. Did you mean to parse the bit in this method and use it when creating the corresponding `Format`?

peddisri replied:
This is a mistake in the patch. The original change is as per below. Will fix this.

byte misc = (byte)scratchDescriptorData.readUnsignedByte();
          boolean isWideAspectRatio = ((misc & 0x60) == 0x60);
          closedCaptionFormats.add(Format.createTextSampleFormat(null, mimeType, null,
              Format.NO_VALUE, 0, language, accessibilityChannel, null,
              isWideAspectRatio));
          scratchDescriptorData.skipBytes(1);

ojw28 commented on library/core/src/main/java/com/google/android/exoplayer2/Format.java:
```diff
@@ -150,6 +150,12 @@
+  /**
+   * Boolean to indicate if the video is wide aspect ratio (16:9) or not (4:3)
+   * Only influences captions if the lines are not middle aligned.
+   * Defaulting to true as sources not having any flags are most probably wide screen
```
At the `Format` level it feels like defaulting to false is more appropriate. If you want to default to true from specific CEA calling sites where `Format` instances are constructed, that would probably be a better place to do it (and put this comment).

I also wonder whether this boolean is the right thing to be adding, or whether something more general would be more flexible (e.g. `float textAspectRatio`). Such a value could also have an unset value (e.g. `-1`), which would allow the decision about what default to assume to be made much later (e.g. in the CEA decoders).

peddisri replied:
CEA spec only indicates if the aspect ratio is for wide screen content or for letter boxed content, does not specify the actual ratio. A widescreen could mean any ratio greater than 1.55 i.e mid way of 1.33 (4:3) to 1.77 (16:9). 
floating point comparisons are tricky. 

https://stackoverflow.com/questions/1088216/whats-wrong-with-using-to-compare-floats-in-java

ojw28 replied:
This class is not specific to CEA, so it's more important to come up with a sufficiently flexible model than to do something that maps directly onto the specific format you're interested in. If the CEA spec only provides "wide" or "not wide" then define two constants somewhere (one for each case), and use those for your comparisons. It would also make sense to default it to `Format.NO_VALUE`.

Alternatively, if you don't think this is ever going to be useful for non-CEA, use the existing `initializationData` field, which is decoder/format specific, and appears to be unused for CEA at the moment. You could define it to contain a single byte[] containing a single byte, which is 1 or 0 based on this flag. That's a much less invasive change, since it lets you revert all changes to this file, and is probably the way to go as I think other caption formats put windowing information in `initializationData` or in the samples already.

As an aside, `accessibilityChannel` should probably be in `initializationData` as well. We might make that change at some point.

peddisri replied:
Ideally, Format in its current form needs to be re-designed anyways because there are many fields that are applicable only to one type of format like Video or audio. In fact even language is applicable only to either audio or text.  

We should re-define Format that just has mimeType and perhaps Id,  and then have derived classes like VideoFormat, AudioFormat, TextFormat etc and have specific fields in each of them. Alternatively, add a list of key/value pairs to make it easily extendable, i.e hash map

I don't know if wide and non-wide captions would be applicable for other subtitle formats or not. I assumed usage of initializationData is like a binary blob that is sent to decoder which is not interpreted at app layer. But by using it in the way you advised, we are now giving meaning to the data elements of the blob. 

So here's what I'm gonna try to do. Introduce a generic hashtable to pass custom parameters. In future you may want to use it to remove accessibilityChannel and any other fields and make the Format Constructor light weight and common for all media types. This will now be similar to the design paradigm of Android's MediaFormat

ojw28 replied:
> I assumed usage of initializationData is like a binary blob that is sent to decoder which is not interpreted at app layer.

It is a binary blob that's sent to the decoder. Where the decoder lives (app layer or platform) is not well defined. Even for audio and video, you can have decoders in both the platform layer (MediaCodec) and the app layer (e.g. the VP9 and FFMPEG extensions). For text the decoders all happen to be in the app layer.

> But by using it in the way you advised, we are now giving meaning to the data elements of the blob.

The blobs have a well defined structure in all cases (else whatever receives it wont know what to do with it). The only distinction is whether the app layer needs to know what the structure is, which it doesn't if it can transparently pass through the blob from the container to the platform in all cases, and which it does if it ever needs to construct or parse one itself.

Defining what we want the blob to look like for CEA formats seems reasonable to me. If it makes it better you can add a `CeaInitializationData` class in the `text.cea` package with methods for converting to/from a `byte[]`, and use that so that the structure of the blob is encapsulated within that class.

> So here's what I'm gonna try to do. Introduce a generic hashtable to pass custom parameters. In future you may want to use it to remove accessibilityChannel and any other fields and make the Format Constructor light weight.

Please don't do this. I agree that `Format` probably needs to be redesigned at some point, but for now `initializationData` is the mechanism for carrying this type of data. Adding a new alternative way of doing things makes things even more confusing.

ojw28 commented on library/core/src/main/java/com/google/android/exoplayer2/Format.java:
```diff
@@ -572,22 +599,23 @@ public static Format createSampleFormat(
         colorInfo, channelCount, sampleRate, pcmEncoding, encoderDelay, encoderPadding,
         selectionFlags, language, accessibilityChannel, subsampleOffsetUs, initializationData,
-        drmInitData, metadata);
+        drmInitData, metadata, true);
```
All of these copy methods need to copy whatever the value of the field is in this instance, rather than hardcoding it to true as part of the copy step.

peddisri replied:
agreed.

peddisri reviewed (commented):
Responded to few comments.

ojw28 commented:
Responded to a couple. I think we need the comments addressed, and in particular all CEA-608 specific changes removed from the text package, before we can make progress here. It would still be preferable if the change could be split into smaller pieces.

ojw28 commented on library/core/src/main/java/com/google/android/exoplayer2/text/Cue.java:
```diff
@@ -98,11 +101,33 @@
+   * The list of pair of cue text and column offsets. Note the {@link CharSequence} may be decorated
+   * with styling spans.
+   */
+  public final List<Pair<Integer, CharSequence>> textRows;
```
Does this effectively duplicate the text in every `Cue`? That doesn't seem ideal from a modelling point of view. Given CEA uses monospace fonts, why can't columns be handled using spaces at the start of each line in `text`?

peddisri replied:
It does not duplicate the text.  Either text field is set or textRows is set, not both. 
Earlier text field used to contain a 4 line text. But now textRows will have 4 elements, each having one line of text and associated column offset. 

In fact, initially we did implement column offset by inserting a simple space. The problem with this approach is that space draws the text background, which does not match the expected reference output when the window color is transparent.

peddisri commented:
Ok I'll address the feedback so far. For the splitting part, let me explore if I can split this into three parts :
1. CEA 608 specific.
2. CEA 708 Decoding/parsing specific
3. CEA 708 Styling Specific (including painter)

Commit:
Support CEA 608/708 captions in MPEG2 TS stream

[Problem] Sarnoff spec complaince tests streams are MPEG2 TS.
Currently, parsing CC data from MPEG2 TS is not supported

[Solution] Parsed CC data from MPEG2 stream because all Sarnoff streams are MPEG2

peddisri commented:
@ojw28  Let's do this pull request in phases. It's hard to split the code into three and I'd like to take baby steps here. The latest pull request incluedes
1. CEA708 decoder changes 
2. Minor chane in CEA608 to support this change.

One you review this, I'll put up another pull request for 608 decoder only, and then lastly styling related.

peddisri commented:
@ojw28  hope you can find some time to review it this week

ojw28 commented on library/core/src/main/java/com/google/android/exoplayer2/C.java:
```diff
@@ -77,6 +77,11 @@ private C() {}
+  /**
+   * The number of 10th of a second per microseconds.
+   */
+  public static final long TENTH_OF_SECOND_PER_MICROSECOND = 100000L;
```
Constant name is wrong way around (it's MICROS_PER_TENTH_OF_SECOND, right?). That aside, this is a pretty weird constant, and it's not likely to be used in many places. Can you just inline the value in the one place it's currently used, or use `C.MICROS_PER_SECOND / 10` there?

peddisri replied:
done

ojw28 commented on library/core/src/main/java/com/google/android/exoplayer2/Format.java:
```diff
@@ -508,7 +533,8 @@ public static Format createSampleFormat(
       @Nullable DrmInitData drmInitData,
-      @Nullable Metadata metadata) {
+      @Nullable Metadata metadata,
+      boolean isWideAspectRatio) {
```
`FormatTest` does not compile after this change

ojw28 commented on library/core/src/main/java/com/google/android/exoplayer2/extractor/ts/DefaultTsPayloadReaderFactory.java:
```diff
@@ -77,8 +77,12 @@ public DefaultTsPayloadReaderFactory(@Flags int flags) {
+      closedCaptionFormats = new ArrayList();
+      closedCaptionFormats.add(Format.createTextSampleFormat(null,
+              MimeTypes.APPLICATION_CEA608,  0, null));
+      closedCaptionFormats.add(Format.createTextSampleFormat(null,
```
This change is causing `TsExtractorTest` to fail. You need to update the sample dump files with the new expected extractor output.

peddisri replied:
done

ojw28 commented on library/core/src/main/java/com/google/android/exoplayer2/Format.java:
```diff
@@ -730,7 +758,8 @@ public boolean equals(@Nullable Object obj) {
         && Arrays.equals(projectionData, other.projectionData)
-        && initializationDataEquals(other);
+        && initializationDataEquals(other)
+        && isWideAspectRatio == other.isWideAspectRatio    ;
```
nit - We have an cheap-to-expensive ordering of comparisons in this method. Please move this one up to the primitive comparisons (also rm whitespace before the semi-colon).

ojw28 commented on library/core/src/main/java/com/google/android/exoplayer2/extractor/ts/H262Reader.java:
```diff
@@ -130,7 +148,25 @@ public void consume(ParsableByteArray data) {
+          bytesAlreadyPassed = -lengthToStartCode;
+        }
+
+        if (userData.endNalUnit(bytesAlreadyPassed)) {
```
In the H264 case when dealing with SEI, we call `NalUnitUtil.unescapeStream` here. Are you certain it's not necessary to do unescaping here also? The equivalent would be something like:
```
int unescapedLength = NalUnitUtil.unescapeStream(userData.nalData, userData.nalLength);
userDataParsable.reset(sei.nalData, unescapedLength);
userDataParsable.setPosition(4); // NAL prefix and nal_unit() header.
userDataReader.consume(sampleTimeUs, userDataParsable);
```

ojw28 replied:
[I have no idea what the answer to this is; I was just comparing the two whilst reviewing]

peddisri replied:
done

ojw28 commented on library/core/src/main/java/com/google/android/exoplayer2/extractor/ts/H262Reader.java:
```diff
@@ -130,7 +148,25 @@ public void consume(ParsableByteArray data) {
+          userDataReader.consume(sampleTimeUs, userDataParsable);
+        }
 
+        if (startCodeValue == START_USER_DATA && data.data[startCodeOffset + 2] == 0x1) {
```
I think you can do `userData.startNalUnit(startCodeValue)` unconditionally. `startNalUnit` already checks the type internally, and I think the data.data check is guaranteed to be true due to `NalUnitUtil.findNalUnit` having found a NAL unit further up in this method.

peddisri replied:
We need to check two bytes : 0x01B2.

ojw28 commented on library/core/src/main/java/com/google/android/exoplayer2/text/SubtitleDecoderFactory.java:
```diff
@@ -107,7 +107,7 @@ public SubtitleDecoder createDecoder(Format format) {
               return new Cea608Decoder(format.sampleMimeType, format.accessibilityChannel);
             case MimeTypes.APPLICATION_CEA708:
-              return new Cea708Decoder(format.accessibilityChannel);
+              return new Cea708Decoder(format.accessibilityChannel, format.isWideAspectRatio);
```
Feel free to just put `isWideAspectRatio` into `initializationData` for now, as suggested in `Format`. However the eventual goal would be to have these decoders just take `initializationData` like the others. We'll probably move `accessibilityChannel` and `packetLength` into `initializationData` in a subsequent change.

(`packetLength` is derived from `sampleMimeType`, which is why `sampleMimeType` is currently being passed to the 608 decoder)

ojw28 reviewed (commented):
I've reviewed the extractor side changes. They look good, thanks, although I've posted some comments. The main one being to revert the `Format` change. The rest are pretty minor.

It would be preferable if you could split this change into the extractor side changes, and the decoder side changes. They're almost completely decoupled, and so there's no reason for them to exist in the same change. We can then get the extractor side changes submitted, and then continue with the decoder side changes as the next step.

peddisri commented:
@ojw28 addressed your comments. split patch into extractor changes first.

ojw28 commented:
Thanks! We'll get this merged next week (most likely finding another way to pass the information through `Format`, but no need to make any changes your side).

peddisri commented:
Thanks. Once you merge it to dev, I'll work on the remaining part.

ojw28 commented on library/core/src/main/java/com/google/android/exoplayer2/extractor/ts/DefaultTsPayloadReaderFactory.java:
```diff
@@ -162,17 +198,24 @@ private SeiReader buildSeiReader(EsInfo esInfo) {
           // Skip easy_reader(1), wide_aspect_ratio(1), reserved(14).
-          scratchDescriptorData.skipBytes(2);
+          byte misc = (byte)scratchDescriptorData.readUnsignedByte();
+          boolean isWideAspectRatio = ((misc & 0x60) == 0x60);
```
`0x60` should be `0x40`, right? (no need to update it here, we can do that as part of our merge).

peddisri replied:
Either of them is okay, because the remaining 6 bits are reserved as 1s anyways as per the spec. But yes, more correct would be 0x40.
